### PR TITLE
fix: use SourceMapSource in js side to support map method

### PR DESCRIPTION
**Pull request by SyMind**
## Summary

The `CompatSource` in Rust is unnecessary and can be simplified. Below are the steps detailing the changes for converting between Rust and JavaScript sources.

1. Rust to JavaScript
    * If the source is a buffer or doesn't have a source map, JavaScript will expose it using `RawSource`.
    * If the source has a source map, JavaScript will expose it using `SourceMapSource`.
    
2. JavaScript to Rust
    * The conversion process remains the same as above.

Fix: #7891

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

**Comment by SyMind** [comment]
!bench

**Comment by SyMind** [comment]
!bench

**Comment by SyMind** [comment]
!bench

**Comment by SyMind** [comment]
!bench

**Comment by SyMind** [comment]
!bench

File: packages/rspack/src/util/source.ts [hunk 1]
```diff
@@ -1,79 +1,49 @@
...
-					return JSON.parse(source.map.toString("utf-8"));
-				}
-
-				return null;
-			}
-		});
+		return new SourceMapSource(
+			source.source,
+			"inmemory://from rust",
+			source.map ? JSON.parse(source.map) : null
```
Comment (h-a-n-a) [inline packages/rspack/src/util/source.ts:17]: Can we lazily call `JSON.parse`?

**Comment by h-a-n-a** [comment]
Most part of this PR is a refactor and due to this PR including a fix, I would recommend you to use 'fix' as the title.

File: packages/rspack/src/util/source.ts [hunk 2]
```diff
@@ -1,79 +1,49 @@
...
-				if (source.map) {
-					return JSON.parse(source.map.toString("utf-8"));
-				}
-
-				return null;
-			}
-		});
+		return new SourceMapSource(
+			source.source,
+			"inmemory://from rust",
```
[the remark this answers is not in the document]
Comment (SyMind) [reply packages/rspack/src/util/source.ts:16]: `filename` is only used when an inner source map exists in `SourceMapSource`.

File: packages/rspack/src/util/source.ts (hunk 1 above)
Comment (SyMind) [reply packages/rspack/src/util/source.ts]: We can. I fix it.

**Review by h-a-n-a — APPROVED** [review]

**Pull request merged**